src/CubedSpheres/cubed_sphere_faces.jl

```diff
@@ -0,0 +1,70 @@
+using Statistics
+using Oceananigans.Architectures
+
+using OffsetArrays: OffsetArray
+
+import Base: getindex, size, show, minimum, maximum
+import Statistics: mean
+
+import Oceananigans.Fields: AbstractField, ReducedField, Field, new_data
+import Oceananigans.BoundaryConditions: FieldBoundaryConditions
+
+struct CubedSphereFaces{E, F}
+    faces :: F
+end
+
+const CubedSphereData = CubedSphereFaces{<:OffsetArray}
+
+const CubedSphereField = AbstractField{X, Y, Z, <:Union{Nothing, AbstractArchitecture}, <:ConformalCubedSphereGrid} where {X, Y, Z}
+const CubedSphereFaceField = AbstractField{X, Y, Z, A, <:ConformalCubedSphereFaceGrid} where {X, Y, Z, A}
+const CubedSphereReducedField = ReducedField{X, Y, Z, A, D, <:ConformalCubedSphereFaceGrid} where {X, Y, Z, A, D}
+
+# There must be a way to dispatch in one function without ambiguity with `new_data.jl`...
+
+function new_data(FT, arch::AbstractCPUArchitecture, grid::ConformalCubedSphereGrid, (X, Y, Z))
+    faces = Tuple(new_data(FT, arch, face_grid, (X, Y, Z)) for face_grid in grid.faces)
+    return CubedSphereFaces{typeof(faces[1]), typeof(faces)}(faces)
+end
+
+function new_data(FT, arch::AbstractGPUArchitecture, grid::ConformalCubedSphereGrid, (X, Y, Z))
+    faces = Tuple(new_data(FT, arch, face_grid, (X, Y, Z)) for face_grid in grid.faces)
+    return CubedSphereFaces{typeof(faces[1]), typeof(faces)}(faces)
+end
+
+function FieldBoundaryConditions(grid::ConformalCubedSphereGrid, (X, Y, Z); user_defined_bcs...)
+
+    faces = Tuple(
+        inject_cubed_sphere_exchange_boundary_conditions(
+            FieldBoundaryConditions(face_grid, (X, Y, Z); user_defined_bcs...),
+            face_number,
+            grid.face_connectivity
+        )
+        for (face_number, face_grid) in enumerate(grid.faces)
+    )
+
+    return CubedSphereFaces{typeof(faces[1]), typeof(faces)}(faces)
+end
+
+#####
+##### Utils
+#####
+
+Base.size(data::CubedSphereData) = (size(data.faces[1])..., length(data.faces))
```

Comment:
I guess this isn't well-defined when the faces have different sizes. Should we check out what `RaggedArray` does to figure out what makes sense? This function could matter when broadcasting, though I think we have to bypass default stuff anyways for cubed sphere so I don't it will come into play...